src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -112,66 +115,66 @@ library GPv2Encoding {
         // effort.
 
         require(
-            encodedOrder.length == ORDER_STRIDE,
-            "GPv2: malformed order data"
+            encodedTrade.length == TRADE_STRIDE,
+            "GPv2: malformed trade data"
         );
 
-        uint8 sellTokenIndex = uint8(encodedOrder[0]);
-        uint8 buyTokenIndex = uint8(encodedOrder[1]);
-        order.sellAmount = abi.decode(encodedOrder[2:], (uint256));
-        order.buyAmount = abi.decode(encodedOrder[34:], (uint256));
+        // NOTE: Solidity allocates, but does not free, memory when:
+        // - initializing the `order` memory pointer
+        // - calling the ABI encoding methods
+        // - calling the `ecrecover` precompile.
+        // However, we can restore the free memory pointer to before we made
+        // allocations to effectively free the memory. This is safe as the
+        // memory used can be discarded, and the memory pointed to by the free
+        // memory pointer **does not have to point to zero-ed out memory**.
+        // <https://solidity.readthedocs.io/en/v0.6.12/internals/layout_in_memory.html>
+        uint256 freeMemoryPointer;
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            freeMemoryPointer := mload(0x40)
+        }
+
+        Order memory order = trade.order;
```

Comment:
So you are saying this allocates memory for as if we were cloning the order into the new variable but behaves like a reference (e.g. we are assigning fields on trade.order)?

You could just access order through the trade object below (e.g. `trade.order.sellAnmount = ...`), but I assume since we are freeing the memory eventually this is not a big issue.

In any case it might be worthwhile raising this behaviour to the solidity team.

Reply:
> So you are saying this allocates memory for as if we were cloning the order

It allocates memory for an uninitialized order (it does not zero out the memory or clone from `trade.order`).

> You could just access order through the trade object below (e.g. trade.order.sellAnmount = ...),

Yes, we can do that as well. The only annoyance is that `trade.order.sellAmount` is now a double `mload`. Its not a big deal at all, just an annoyance. The reason I declared the `Order` here was to make the assembly for computing the order struct hash easier to understand. Otherwise, we would have to get the memory offset for the order ourselves.

```solidity
assembly {
    let order := mload(trade) // `order` field is the first field of `Trade` struct
    orderDigest := keccak256(order, 320)
}
```

Reply:
> In any case it might be worthwhile raising this behaviour to the solidity team.

I'll first ask if anyone knows of a better way to do this, I might just not have been able to find this online.

Reply:
Looks like the Solidity team just fixed the issue :tada: https://github.com/ethereum/solidity/pull/10341

Reply:
Wow, what a coincidence! What does it mean for us? Would we consider switching to a freshly baked Solidity 0.7.6?

Reply:
Hehe it's not a coincidence, I reached out, they just fixed it ridiculously fast.